### DISCUSSION TITLE
Ejbca Database Description

### DISCUSSION THREAD

**Discussion by TuanTAg** [Keyfactor/ejbca-ce#183] [Q&A] [answered]
Hi,
Ejbca is good application, I am trying to understand about database of Ejbca.
Is there any link describe about Ejbca Database? about tables, columns,...?
Thank you!

**Comment by mike-agrenius-kushner** [comment, marked as answer]
Yeah, sure – you have the complete ORM description for all supported database vendors here: https://github.com/Keyfactor/ejbca-ce/tree/main/modules/ejbca-entity/resources

Cheers,
Mike

**Comment by TuanTAg** [comment]
Thank so much @mike-agrenius-kushner

**Comment by TuanTAg** [comment]
Hi @mike-agrenius-kushner,
Is there document can help understand what purpose of tables?
Example: what purpose of table AcmeAccountData, ACMECHALLENGEDAT?

**Reply by mike-agrenius-kushner** [reply]
Not without getting into a much deeper discussion about the inner workings of EJBCA – if you want to find out what each table does then I suggest digging into the code and reading the documentation on the corresponding Data Session Beans. Note that some of the tables are only referenced in EE, but we don't carry separate database definitions for CE.  If you have any specific questions I'll gladly help you out. 

Cheers,
Mike

**Reply by TuanTAg** [reply]
Thanks Mike!

**Reply by StocktonC** [reply]
I also looking for tables meaning, field limitation, list of available value of each field etc.
Say, for CertificateData table, i need to guess that status=20 stands for valid, status=40 stands for revoked.
It would be give help for program development if i want to insert a record into CertificateData table after key generated, which field(s) (value, syntax, and format) is required for OCSP verification.

**Reply by primetomas** [reply]
If you look in the orm file, for example for MySQL: https://github.com/Keyfactor/ejbca-ce/blob/main/modules/ejbca-entity/resources/orm-ejbca-mysql.xml
You will see that the class is: class="org.cesecore.certificates.certificate.CertificateData
If you look into this file you will find the fields you ask about: https://github.com/Keyfactor/ejbca-ce/blob/main/modules/cesecore-entity/src/org/cesecore/certificates/certificate/CertificateData.java
The javadoc in that class points to CertificateConstants: https://github.com/Keyfactor/ejbca-ce/blob/main/modules/cesecore-common/src/org/cesecore/certificates/certificate/CertificateConstants.java

Cheers,
Tomas
